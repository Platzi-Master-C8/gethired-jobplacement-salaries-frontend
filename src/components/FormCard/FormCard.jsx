import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Typography, Card, Chip, Box } from '@mui/material';

import Select from 'Components/Commons/Select';

import { selectTechnologies, selectJobs } from 'App/ListData/selectors';

import { fetchListData } from 'App/ListData/slice';

import { getListByName } from 'Services/salaries';
import { selectSeniority } from '../../app/ListData/selectors';

const FormCard = ({
    onChange,
    title,
    values,
    listTechnologies,
    listJobs,
    ListSenority,
    children,
    onDelete,
    addListData,
}) => {
    const { title_id, technologies, seniority, english_level } = values;
    // const [ListSenority, setListSenority] = useState([]);
    const [ListEnglish, setListEnglish] = useState([]);

    useEffect(() => {
        addListData();
        // setListSenority(getListByName('senority'));
        setListEnglish(getListByName('english'));
    }, [addListData]);

    return (
        <Card sx={{ p: 2, boxShadow: 3, mt: 2 }}>
            {!!title && (
                <Typography sx={{ mb: 2 }} variant="h2">
                    {title}
                </Typography>
            )}
            <Select
                label="Job Title"
                value={title_id}
                onChange={onChange}
                id="label-job"
                name="title_id"
                options={listJobs}
            />
            <Select
                label="Technologies"
                value={technologies}
                onChange={onChange}
                id="label-technologies"
                name="technologies"
                options={listTechnologies}
                multiple
                renderValue={(selected) => (
                    <Box sx={{ display: 'flex', flexWrap: 'wrap', gap: 0.5 }}>
                        {selected.map((value) => (
                            <Chip
                                key={value}
                                label={value}
                                variant="outlined"
                                color="primary"
                                onDelete={(event) => onDelete(event, value, 'technologies')}
                                onMouseDown={(event) => event.stopPropagation()}
                            />
                        ))}
                    </Box>
                )}
            />
            <Select
                label="Seniority"
                value={seniority}
                onChange={onChange}
                id="label-seniority"
                name="seniority"
                options={ListSenority}
            />
            <Select
                label="English Level"
                value={english_level}
                onChange={onChange}
                id="label-englishLevel"
                name="english_level"
                options={ListEnglish}
            />
            {children}
        </Card>
    );
};

FormCard.propTypes = {
    values: PropTypes.shape({
        title_id: PropTypes.string,
        technologies: PropTypes.arrayOf(PropTypes.string),
        seniority: PropTypes.string,
        english_level: PropTypes.string,
    }).isRequired,
    children: PropTypes.node,
    listTechnologies: PropTypes.arrayOf(PropTypes.string).isRequired,
    listJobs: PropTypes.arrayOf(PropTypes.string).isRequired,
    ListSenority: PropTypes.arrayOf(PropTypes.string).isRequired,
    onChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired,
    title: PropTypes.string,
    addListData: PropTypes.func.isRequired,
};

FormCard.defaultProps = {
    title: null,
    children: null,
};

const mapStateToProps = (state) => ({
    listTechnologies: selectTechnologies(state),
    listJobs: selectJobs(state),
    ListSenority: selectSeniority(state),
});

const mapDispatchToProps = (dispatch) => ({
    addListData: () => dispatch(fetchListData()),
});

export default connect(mapStateToProps, mapDispatchToProps)(FormCard);
