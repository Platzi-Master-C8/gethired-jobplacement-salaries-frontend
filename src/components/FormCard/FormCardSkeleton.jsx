import React, { Fragment } from 'react';

import Skeleton from '@mui/material/Skeleton';

const FormCardSkeleton = () => {
    return (
        <Fragment>
            <Skeleton sx={{ mb: 3, ml: 4 }} variant="rectangular" height={56} />
            <Skeleton sx={{ mb: 3, ml: 4 }} variant="rectangular" height={56} />
            <Skeleton sx={{ mb: 3, ml: 4 }} variant="rectangular" height={56} />
            <Skeleton sx={{ mb: 3, ml: 4 }} variant="rectangular" height={56} />
            <Skeleton sx={{ mb: 3, ml: 4 }} variant="rectangular" height={56} />
            <Skeleton sx={{ mb: 3, ml: 4 }} variant="rectangular" height={56} />
        </Fragment>
    );
};

export default FormCardSkeleton;
