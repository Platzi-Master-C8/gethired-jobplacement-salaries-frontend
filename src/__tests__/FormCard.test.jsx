import React from 'react';
import { render, waitFor } from '@testing-library/react';

import FormCard from 'Components/FormCard';

const initialValues = {
    jobTitle: '',
    technologies: [],
    senority: '',
    englishLevel: '',
};

describe('Tests for component <FormCard/>', () => {
    const handleChange = jest.fn();
    test('Take Snapshot', () => {
        const { asFragment } = render(
            <FormCard title="Calculate Salary" values={initialValues} onChange={handleChange} />,
        );

        expect(asFragment()).toMatchSnapshot();
    });
    test('Expect title Calculate Salary for card', async () => {
        const { getByText } = render(
            <FormCard title="Calculate Salary" values={initialValues} onChange={handleChange} />,
        );
        await waitFor(() => expect(getByText(/Calculate Salary/i)).toBeInTheDocument());
    });
    test('Have 4 selects', () => {
        const { container } = render(<FormCard values={initialValues} onChange={handleChange} />);
        expect(container.querySelectorAll('input')).toHaveLength(4);
    });
});
