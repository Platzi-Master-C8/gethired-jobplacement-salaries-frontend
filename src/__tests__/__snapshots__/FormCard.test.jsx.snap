// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for component <FormCard/> Take Snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-r1w9s9-MuiPaper-root-MuiCard-root"
  >
    <h2
      class="MuiTypography-root MuiTypography-h2 css-vqgaav-MuiTypography-root"
    >
      Calculate Salary
    </h2>
    <div
      aria-expanded="false"
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1kpj6rr-MuiAutocomplete-root"
      role="combobox"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-1"
          id="mui-1-label"
        >
          Job Title
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1w8yruj-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiFilledInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1gctnaj-MuiInputBase-input-MuiFilledInput-input"
            id="mui-1"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-expanded="false"
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1kpj6rr-MuiAutocomplete-root"
      role="combobox"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="mui-2"
          id="mui-2-label"
        >
          Technologies
        </label>
        <div
          class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1w8yruj-MuiInputBase-root-MuiFilledInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiFilledInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1gctnaj-MuiInputBase-input-MuiFilledInput-input"
            id="mui-2"
            spellcheck="false"
            type="text"
            value=""
          />
          <div
            class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-3lxuvf-MuiFormControl-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        id="label-seniority"
      >
        Seniority
      </label>
      <div
        class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-cio0x1-MuiInputBase-root-MuiFilledInput-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="label-seniority mui-component-select-seniority"
          class="MuiSelect-select MuiSelect-filled MuiFilledInput-input MuiInputBase-input css-d9oaum-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
          id="mui-component-select-seniority"
          role="button"
          tabindex="0"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          name="seniority"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth css-3lxuvf-MuiFormControl-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary css-e4w4as-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        id="label-englishLevel"
      >
        English Level
      </label>
      <div
        class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-cio0x1-MuiInputBase-root-MuiFilledInput-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="label-englishLevel mui-component-select-english_level"
          class="MuiSelect-select MuiSelect-filled MuiFilledInput-input MuiInputBase-input css-d9oaum-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
          id="mui-component-select-english_level"
          role="button"
          tabindex="0"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          name="english_level"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
